File: headless_harness_datagen/harness/chakra/task_games_08/entity.py
# entity.py
"""Entity definitions for Deepvault Survey roguelike.
Includes Player and three enemy types with specified behaviors.
"""
import random
from typing import Tuple, List

# Helper for Manhattan distance
def manhattan(a: Tuple[int, int], b: Tuple[int, int]) -> int:
    return abs(a[0] - b[0]) + abs(a[1] - b[1])

class Player:
    def __init__(self, start_pos: Tuple[int, int], prng_func):
        self.pos = start_pos
        self.hp = 10
        self.max_hp = 10
        self.attack_damage = 2
        self.prng = prng_func
        # inventory managed separately

    def move(self, dx: int, dy: int, dungeon):
        x, y = self.pos
        nx, ny = x + dx, y + dy
        # bounds check
        if 0 <= nx < dungeon.width and 0 <= ny < dungeon.height:
            if dungeon.map[ny][nx] != "#":
                self.pos = (nx, ny)

class BaseEnemy:
    def __init__(self, glyph: str, hp: int, dmg: int, color: Tuple[int, int, int]):
        self.glyph = glyph
        self.hp = hp
        self.dmg = dmg
        self.color = color
        self.pos: Tuple[int, int] = (0, 0)
        self.state = {}

    def set_position(self, pos: Tuple[int, int]):
        self.pos = pos

    def take_turn(self, player: Player, dungeon, prng_func, entities: List['BaseEnemy']):
        pass

class RustHusk(BaseEnemy):
    """Wanders until player within 5 tiles, then chases via simple BFS step towards player."""
    def __init__(self, pos: Tuple[int, int]):
        super().__init__(glyph='h', hp=4, dmg=1, color=(200, 100, 0))
        self.pos = pos
        self.aware = False

    def take_turn(self, player, dungeon, prng_func, entities):
        if not self.aware:
            if manhattan(self.pos, player.pos) <= 5:
                self.aware = True
        if self.aware:
            # simple greedy step towards player avoiding walls
            px, py = player.pos
            x, y = self.pos
            dx = 1 if px > x else -1 if px < x else 0
            dy = 1 if py > y else -1 if py < y else 0
            # try horizontal move first
            if dx != 0 and dungeon.map[y][x + dx] != '#':
                self.pos = (x + dx, y)
            elif dy != 0 and dungeon.map[y + dy][x] != '#':
                self.pos = (x, y + dy)
        else:
            # wander randomly
            dirs = [(1,0),(-1,0),(0,1),(0,-1)]
            random.shuffle(dirs)
            for dx, dy in dirs:
                nx, ny = self.pos[0] + dx, self.pos[1] + dy
                if 0 <= nx < dungeon.width and 0 <= ny < dungeon.height and dungeon.map[ny][nx] != '#':
                    self.pos = (nx, ny)
                    break
        # attack if adjacent
        if manhattan(self.pos, player.pos) == 1:
            player.hp -= self.dmg

class SentryCoil(BaseEnemy):
    """Immobile; if player shares row/col with clear LOS, telegraphs then zaps next turn."""
    def __init__(self, pos: Tuple[int, int]):
        super().__init__(glyph='s', hp=3, dmg=2, color=(0, 0, 255))
        self.pos = pos
        self.charge = 0  # 0 = idle, 1 = telegraphing

    def line_of_sight(self, player, dungeon):
        px, py = player.pos
        x, y = self.pos
        if x == px:
            step = 1 if py > y else -1
            for ny in range(y + step, py, step):
                if dungeon.map[ny][x] == '#':
                    return False
            return True
        if y == py:
            step = 1 if px > x else -1
            for nx in range(x + step, px, step):
                if dungeon.map[y][nx] == '#':
                    return False
            return True
        return False

    def take_turn(self, player, dungeon, prng_func, entities):
        if self.charge == 1:
            # zap now
            if self.line_of_sight(player, dungeon):
                player.hp -= self.dmg
            self.charge = 0
            return
        if self.line_of_sight(player, dungeon):
            # start telegraph
            self.charge = 1
            # UI can show flashing via state, ignored here

class ScavRat(BaseEnemy):
    """Wanders; consumes items on tile; attacks only when adjacent."""
    def __init__(self, pos: Tuple[int, int]):
        super().__init__(glyph='r', hp=2, dmg=1, color=(150, 75, 0))
        self.pos = pos

    def take_turn(self, player, dungeon, prng_func, entities):
        # wander randomly
        dirs = [(1,0),(-1,0),(0,1),(0,-1)]
        random.shuffle(dirs)
        for dx, dy in dirs:
            nx, ny = self.pos[0] + dx, self.pos[1] + dy
            if 0 <= nx < dungeon.width and 0 <= ny < dungeon.height and dungeon.map[ny][nx] != '#':
                self.pos = (nx, ny)
                break
        # attack if adjacent
        if manhattan(self.pos, player.pos) == 1:
            player.hp -= self.dmg

File: headless_harness_datagen/harness/chakra/task_games_08/web/src/entity.ts
/** Player + enemies for Deepvault Survey. */

export type Pos = [number, number];

function manhattan(a: Pos, b: Pos): number {
  return Math.abs(a[0] - b[0]) + Math.abs(a[1] - b[1]);
}

export class Player {
  public pos: Pos;
  public hp = 10;
  public maxHp = 10;
  public attackDamage = 2;
  constructor(startPos: Pos) {
    this.pos = [...startPos] as Pos;
  }
}

export abstract class BaseEnemy {
  public pos: Pos;
  public hp: number;
  public dmg: number;
  public glyph: string;
  public color: string;
  constructor(pos: Pos, glyph: string, hp: number, dmg: number, color: string) {
    this.pos = [...pos] as Pos;
    this.glyph = glyph;
    this.hp = hp;
    this.dmg = dmg;
    this.color = color;
  }
  abstract takeTurn(
    player: Player,
    dungeon: { map: string[][]; width: number; height: number; walkable: (x: number, y: number) => boolean },
    prng: () => number
  ): void;
}

function tryStep(
  pos: Pos,
  dx: number,
  dy: number,
  dungeon: { walkable: (x: number, y: number) => boolean }
): Pos {
  const nx = pos[0] + dx;
  const ny = pos[1] + dy;
  if (dungeon.walkable(nx, ny)) return [nx, ny];
  return pos;
}

export class RustHusk extends BaseEnemy {
  aware = false;
  constructor(pos: Pos, hpScale = 0) {
    super(pos, "h", 4 + hpScale, 1, "#ff8800");
  }
  takeTurn(player, dungeon, prng) {
    if (!this.aware && manhattan(this.pos, player.pos) <= 5) this.aware = true;
    if (this.aware) {
      const dx = Math.sign(player.pos[0] - this.pos[0]);
      const dy = Math.sign(player.pos[1] - this.pos[1]);
      if (dx) this.pos = tryStep(this.pos, dx, 0, dungeon);
      else if (dy) this.pos = tryStep(this.pos, 0, dy, dungeon);
    } else {
      const dirs: Pos[] = [
        [1, 0],
        [-1, 0],
        [0, 1],
        [0, -1],
      ];
      const [dx, dy] = dirs[Math.floor(prng() * 4)];
      this.pos = tryStep(this.pos, dx, dy, dungeon);
    }
    if (manhattan(this.pos, player.pos) === 0 || manhattan(this.pos, player.pos) === 1) {
      // if on same tile after move, step back conceptually — attack adjacent
    }
    if (manhattan(this.pos, player.pos) === 1) player.hp -= this.dmg;
  }
}

export class SentryCoil extends BaseEnemy {
  charge = 0;
  telegraphing = false;
  constructor(pos: Pos, hpScale = 0) {
    super(pos, "s", 3 + hpScale, 2, "#00aaff");
  }
  private hasLOS(player: Player, dungeon: { map: string[][] }): boolean {
    const [px, py] = player.pos;
    const [sx, sy] = this.pos;
    if (px === sx) {
      const step = py > sy ? 1 : -1;
      for (let y = sy + step; y !== py; y += step) if (dungeon.map[y][sx] === "#") return false;
      return true;
    }
    if (py === sy) {
      const step = px > sx ? 1 : -1;
      for (let x = sx + step; x !== px; x += step) if (dungeon.map[sy][x] === "#") return false;
      return true;
    }
    return false;
  }
  takeTurn(player, dungeon, _prng) {
    this.telegraphing = false;
    if (this.charge === 1) {
      if (this.hasLOS(player, dungeon)) player.hp -= this.dmg;
      this.charge = 0;
      return;
    }
    if (this.hasLOS(player, dungeon)) {
      this.charge = 1;
      this.telegraphing = true;
    }
  }
}

export class ScavRat extends BaseEnemy {
  constructor(pos: Pos, hpScale = 0) {
    super(pos, "r", 2 + hpScale, 1, "#aa5500");
  }
  takeTurn(player, dungeon, prng) {
    const dirs: Pos[] = [
      [1, 0],
      [-1, 0],
      [0, 1],
      [0, -1],
    ];
    const [dx, dy] = dirs[Math.floor(prng() * 4)];
    this.pos = tryStep(this.pos, dx, dy, dungeon);
    if (manhattan(this.pos, player.pos) === 1) player.hp -= this.dmg;
  }
}
